tt-train/sources/ttml/core/indestructible.hpp

```diff
@@ -0,0 +1,38 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+#include <type_traits>
+#include <utility>
+
+namespace ttml::core {
+
+template <typename T>
+class Indestructible {
```

Comment:
I think this is a great library, but perhaps we can use non-static storage for the underlying object, so that the use cases aren't bound to singletons? The caller may then choose to create a static variable (like you did for `Autocontext`):
```
static Indestructible<AutoContext> context;
// context dtor will never run...
```

Reply:
Agree, probably I can do it better. Ill think about it. Currently it is a workaround to avoid calling close device too late.